File: dff/api/datatype/magichandler.py
# DFF -- An Open Source Digital Forensics Framework
# Copyright (C) 2009-2013 ArxSys
# This program is free software, distributed under the terms of
# the GNU General Public License Version 2. See the LICENSE file
# at the top of the source tree.
#  
# See http://www.digital-forensic.org for more information about this
# project. Please do not directly contact any of the maintainers of
# DFF for assistance; the project provides a web site, mailing lists
# and IRC channels for your use.
# 
# Author(s):
#  Solal J. <sja@digital-forensic.org>
#
import os, sys

if hasattr(sys, "frozen"):
   from dff.api.magic import magic
else:
#    try:
#       import magic
#    except:
   from dff.api.magic import magic

from dff.api.exceptions.libexceptions import vfsError 
from dff.api.types.libtypes import Variant
from dff.api.datatype.libdatatype import DataTypeHandler, DataTypeManager

class Magic(DataTypeHandler):
  def __init__(self):
     DataTypeHandler.__init__(self)
     self.__disown__()
     self.mgc_path = None
     if hasattr(sys, "frozen"):
        self.mgc_path = os.path.abspath(os.path.join(os.path.dirname(sys.executable), "resources/magic.mgc"))
     else:
        mgcpath = os.path.realpath(__file__)
        idx = mgcpath.rfind("api")
        if idx != -1:
           mgcpath = mgcpath[:idx+3]
        self.mgc_path = os.path.join(mgcpath, "magic", "magic.mgc")


  def typeFromBuffer(self, buff):
     if not len(buff):
        return "empty"
     mime = None
     buffmime = "unknown"
     f = None
     try:
        mime = magic.open(magic.MAGIC_NONE)
        mime.load(self.mgc_path)
        buffmime = mime.buffer(buff)
     except Exception as e:
        print e
        #print "Magic error can't read buffer on node : ",  node.absolute(), "\n", e
        buffmime = "error"
     finally:
        if mime is not None:
           mime.close()	
     return buffmime
     

  def type(self, node):
    if node.size() > 0:
       mime = None
       filemime = "unknown"
       f = None
       try:
          mime = magic.open(magic.MAGIC_NONE)
          mime.load(self.mgc_path)
          f = node.open()
          #cannot read less than 0x2000 because of vshadow signature starting @0x1e00
          buff = f.read(0x2000)
          filemime = mime.buffer(buff)
       except Exception as e:
          #print "Magic error can't read buffer on node : ",  node.absolute(), "\n", e
          filemime = "error"
       finally:
          if f is not None:
            f.close()
          if mime is not None:
             mime.close()	
       return filemime
    elif node.hasChildren() or node.isDir():
       return "directory"
    else:
       return "empty"

magicHandler = Magic()

File: dff/api/magic/magichandler.cpp
/*
 * DFF -- An Open Source Digital Forensics Framework
 * Copyright (C) 2009-2013 ArxSys
 * This program is free software, distributed under the terms of
 * the GNU General Public License Version 2. See the LICENSE file
 * at the top of the source tree.
 *  
 * See http://www.digital-forensic.org for more information about this
 * project. Please do not directly contact any of the maintainers of
 * DFF for assistance; the project provides a web site, mailing lists
 * and IRC channels for your use.
 * 
 * Author(s):
 *  Frederic B. <fba@digital-forensic.org>
 */

#include "magichandler.hpp"
#include "vfile.hpp"
#include "exceptions.hpp"

#include <magic.h>
//#include "file.h"



MagicHandler::MagicHandler() : DataTypeHandler("magic")
{
}


MagicHandler::~MagicHandler()
{
}


std::string	MagicHandler::type(Node* node)
{
  VFile*	vf;
  void*		buffer;
  int32_t	rbytes;
  std::string	result;

  vf = NULL;
  buffer = NULL;
  result = std::string("empty");
  if (node != NULL)
    {
      try
	{
	  if ((vf = node->open()) != NULL)
	    {
	      if ((buffer = malloc(4096 * sizeof(uint8_t))) != NULL)
		{
		  rbytes = vf->read(buffer, 4096);
	  	  vf->close();
	  	  //result = this->__magic(buffer, rbytes);
		}
	    }
	}
      catch (vfsError)
	{
	}
    }
  if (buffer != NULL)
    free(buffer);
  return result;
}


std::string	MagicHandler::__magic(void* buffer, uint32_t size)
{
  magic_t	ms;
  const char*	tmp;
  std::string	result;

  //result = std::string("empty");
  ms = NULL;
  tmp = NULL;
  if ((ms = magic_open(MAGIC_CONTINUE)) != NULL)
    {
      if (magic_load(ms, "/home/udgover/projects/magic-custom/file-orig/magic/magic.mgc") != -1)	
       	{
      	  if ((tmp = magic_buffer(ms, buffer, size)) != NULL)
      	    result = std::string(tmp);
      	}
      magic_close(ms);
    }
  return result;
}


// std::string	MagicHandler::__magic(void* buffer, uint32_t size)
// {
//   struct magic_set*	ms;
//   const char*		tmp;
//   std::string		result;
  
//   ms = NULL;
//   tmp = NULL;
//   result = std::string("empty");
//   if ((ms = file_ms_alloc(MAGIC_CONTINUE)) != NULL)
//     {
//       if (file_apprentice(ms, "/home/udgover/projects/magic-custom/file-orig/magic/magic.mgc", FILE_LOAD) != 1)
// 	{
// 	  if (file_reset(ms) == -1)
// 	    return result;
// 	  if (file_buffer(ms, -1, NULL, buffer, size) == -1)
// 	    return result;
// 	  if ((tmp = file_getbuffer(ms)) != NULL)
// 	    result = std::string(tmp);
// 	}
//       file_ms_free(ms);
//     }
//   return result;
// }


static MagicHandler* m = new MagicHandler();
